backup/protocol.py

```diff
@@ -0,0 +1,70 @@
+'''
+Socket-based remote backup protocol. This is used to create a connection to a backup backend, and send5 it incremental database updates.
+'''
+import socket
+import struct
+from typing import Tuple
+import zlib
+
+from backend import Change
+
+class PacketType:
+    CHANGE = 0x01
+    SNAPSHOT = 0x02
+    REWIND = 0x03
+    REQ_METADATA = 0x04
+    RESTORE = 0x05
+    ACK = 0x06
+    NACK = 0x07
+    METADATA = 0x08
+    DONE = 0x09
+    COMPACT = 0x0a
+    COMPACT_RES = 0x0b
+
+PKT_CHANGE_TYPES = {PacketType.CHANGE, PacketType.SNAPSHOT}
+
+def recvall(sock: socket.socket, n: int) -> bytearray:
+    '''Receive exactly n bytes from a socket.'''
+    buf = bytearray(n)
+    view = memoryview(buf)
+    ptr = 0
+    while ptr < n:
+        count = sock.recv_into(view[ptr:])
+        if count == 0:
+            raise IOError('Premature end of stream')
+        ptr += count
+    return buf
+
+def send_packet(sock: socket.socket, typ: int, payload: bytes) -> None:
+    sock.sendall(struct.pack('!BI', typ, len(payload)))
+    sock.sendall(payload)
+
+def recv_packet(sock: socket.socket) -> Tuple[int, bytes]:
+    (typ, length) = struct.unpack('!BI', recvall(sock, 5))
+    payload = recvall(sock, length)
+    return (typ, payload)
+
+def change_from_packet(typ, payload):
+    '''Convert a network packet to a Change object.'''
+    if typ == PacketType.CHANGE:
+        (version, ) = struct.unpack('!I', payload[0:4])
+        payload = zlib.decompress(payload[4:])
```

Comment:
Are all changes zlib compressed? I was considering adding compression as well, and it might make sense to push that down to the file-backend as well, i.e., have a common type structure that allows for compression to be orthogonal to the message type.

Something like using the top nibble to describe the encoding and compression, while using the bottom nibble to be the type Snapshot or Change.

wdyt?

Reply:
I assume that the protocol is fully trusted at this point, so decompression bombs are out of scope :-)

Reply:
Yes, all changes and snapshots are compressed. And yes, the protocol is trusted, both from the client and server side, I don't think you can do much else in a backup protocol.
(Authentication is explicitly out of scope)

Reply:
Sounds good to me :+1: Let's bikeshed in a new issue/PR, and get this merged asap :-)